File: Submariner/SBTracklistController.m
//
//  SBTracklistController.m
//  Submariner
//
//  Created by Rafaël Warnault on 06/06/11.
//
//  Copyright (c) 2011-2014, Rafaël Warnault
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  * Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  * Neither the name of the Read-Write.fr nor the names of its
//  contributors may be used to endorse or promote products derived from
//  this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
//  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
//  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
//  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#import "SBTracklistController.h"
#import "SBDatabaseController.h"

#import "Submariner-Swift.h"




@interface SBTracklistController ()
- (void)playerPlaylistUpdatedNotification:(NSNotification *)notification;
@end



@implementation SBTracklistController

+ (NSString *)nibName {
    return @"Tracklist";
}


- (NSString*)title {
    return @"Tracklist";
}


@synthesize databaseController;



- (void)dealloc {
    // remove player observer
    [[NSNotificationCenter defaultCenter] removeObserver:self name: @"SBPlayerPlaylistUpdatedNotification" object:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self forKeyPath:@"playlist"];
}



- (void)loadView {
    [super loadView];
    
    [playlistTableView setTarget:self];
    [playlistTableView setDoubleAction:@selector(trackDoubleClick:)];
    [playlistTableView registerForDraggedTypes:[NSArray arrayWithObjects:SBTracklistTableViewDataType, SBLibraryTableViewDataType, nil]];
    
    // observer playlist change
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(playerPlaylistUpdatedNotification:)
                                                 name:@"SBPlayerPlaylistUpdatedNotification"
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                           forKeyPath:@"playlist"  
                                              options:(NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld) 
                                              context:(__bridge void*)[SBPlayer sharedInstance]];
}





#pragma mark -
#pragma mark Observers

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    if([keyPath isEqualToString:@"playlist"]) {
        [playlistTableView reloadData];
    }
}





#pragma mark -
#pragma mark IBActions

- (IBAction)trackDoubleClick:(id)sender {
    NSInteger selectedRow = [playlistTableView selectedRow];
    if(selectedRow != -1) {
        SBTrack *clickedTrack = [[[SBPlayer sharedInstance] playlist] objectAtIndex:selectedRow];
        if(clickedTrack) {
            
            // stop current playing tracks
            [[SBPlayer sharedInstance] stop];
            
            // play track
            [[SBPlayer sharedInstance] playTrack:clickedTrack];
        }
    }
}




- (IBAction)removeTrack:(id)sender {
    NSInteger selectedRow = [playlistTableView selectedRow];
    
    if(selectedRow != -1) {
        [[SBPlayer sharedInstance] removeTrackIndexSet: [playlistTableView selectedRowIndexes]];
        [playlistTableView reloadData];
    }
}


- (IBAction)cleanTracklist:(id)sender {
    [[SBPlayer sharedInstance] clear];
    [playlistTableView reloadData];
}


- (IBAction)closeTracklist:(id)sender {
    [databaseController toggleTrackList:sender];
}


- (IBAction)delete: (id)sender {
    [self removeTrack: sender];
}


- (IBAction)showSelectedInFinder:(id)sender {
    NSInteger selectedRow = [playlistTableView selectedRow];
    
    if(selectedRow == -1) {
        return;
    }
    
    [self showTracksInFinder: [[SBPlayer sharedInstance] playlist] selectedIndices: playlistTableView.selectedRowIndexes];
}


- (IBAction)showSelectedInLibrary:(id)sender {
    NSInteger selectedRow = [playlistTableView selectedRow];
    
    if(selectedRow == -1) {
        return;
    }
    
    // only makes sense to have a single track, imho
    NSUInteger index = playlistTableView.selectedRowIndexes.firstIndex;
    SBTrack *track = (SBTrack*)[[[SBPlayer sharedInstance] playlist] objectAtIndex: index];
    [[self databaseController] goToTrack: track];
}


- (IBAction)downloadSelected:(id)sender {
    NSInteger selectedRow = [playlistTableView selectedRow];
    
    if(selectedRow != -1) {
        [self downloadTracks: [[SBPlayer sharedInstance] playlist] selectedIndices: playlistTableView.selectedRowIndexes databaseController: databaseController];
    }
}


- (IBAction)createNewLocalPlaylistWithSelectedTracks:(id)sender {
    NSInteger selectedRow = [playlistTableView selectedRow];
    
    if(selectedRow == -1) {
        return;
    }
    
    [self createLocalPlaylistWithSelected: [[SBPlayer sharedInstance] playlist] selectedIndices: playlistTableView.selectedRowIndexes databaseController: self.databaseController];
}


#pragma mark -
#pragma mark Player Notifications

- (void)playerPlaylistUpdatedNotification:(NSNotification *)notification {

    [playlistTableView reloadData];
}





#pragma mark -
#pragma mark NSTableView DataSource

- (NSInteger)numberOfRowsInTableView:(NSTableView *)tableView {
    return [[[SBPlayer sharedInstance] playlist] count];
}

- (id)tableView:(NSTableView *)tableView objectValueForTableColumn:(NSTableColumn *)tableColumn row:(NSInteger)row {
    id value = nil;
    
    if([[tableColumn identifier] isEqualToString:@"isPlaying"]) {
        SBTrack *track = (SBTrack *)[[[SBPlayer sharedInstance] playlist] objectAtIndex:row];
        if([[track isPlaying] boolValue]) {
            value = [NSImage imageWithSystemSymbolName: @"speaker.fill" accessibilityDescription: @"Playing"];
        }
    }
    if([[tableColumn identifier] isEqualToString:@"title"])
        value = [[[[SBPlayer sharedInstance] playlist] objectAtIndex:row] itemName];
    
    if([[tableColumn identifier] isEqualToString:@"artist"]) {
        SBTrack *track = [[[SBPlayer sharedInstance] playlist] objectAtIndex:row];
        if (track.artistName == nil || [track.artistName isEqualToString: @""]) {
            value = track.album.artist.itemName;
        } else {
            value = track.artistName;
        }
    }
    
    if([[tableColumn identifier] isEqualToString:@"duration"])
        value = [[[[SBPlayer sharedInstance] playlist] objectAtIndex:row] durationString];
    
    if([[tableColumn identifier] isEqualToString:@"online"]) {
        SBTrack *track = (SBTrack *)[[[SBPlayer sharedInstance] playlist] objectAtIndex:row];
        if (track.localTrack != nil || track.isLocal.boolValue == YES) {
            value = [NSImage imageWithSystemSymbolName: @"bolt.horizontal.fill" accessibilityDescription: @"Cached"];
        } else {
            value = [NSImage imageWithSystemSymbolName: @"bolt.horizontal" accessibilityDescription: @"Online"];
        }
    }
    return value;
}




#pragma mark -
#pragma mark NSTableView Delegate

- (BOOL)tableView:(NSTableView *)tv writeRowsWithIndexes:(NSIndexSet *)rowIndexes toPasteboard:(NSPasteboard*)pboard
{
    // internal drop track
    NSError *error = nil;
    NSData *data = [NSKeyedArchiver archivedDataWithRootObject:rowIndexes requiringSecureCoding: YES error: &error];
    if (error != nil) {
        NSLog(@"Error archiving track URIs: %@", error);
        return NO;
    }
    [pboard declareTypes:[NSArray arrayWithObject:SBTracklistTableViewDataType] owner:self];
    [pboard setData:data forType:SBTracklistTableViewDataType];
    
    return YES;
}


- (NSDragOperation)tableView:(NSTableView*)tv validateDrop:(id <NSDraggingInfo>)info proposedRow:(NSInteger)row proposedDropOperation:(NSTableViewDropOperation)op
{
    
    if(row == -1)
        return NSDragOperationNone;
    
    if(op == NSTableViewDropAbove) {
        // internal drop track
        if ([[[info draggingPasteboard] types] containsObject:SBTracklistTableViewDataType] || [[[info draggingPasteboard] types] containsObject:SBLibraryTableViewDataType] ) {
            return NSDragOperationMove;
        }
    }
    
    return NSDragOperationNone;
}


- (BOOL)tableView:(NSTableView *)aTableView acceptDrop:(id <NSDraggingInfo>)info
              row:(NSInteger)row dropOperation:(NSTableViewDropOperation)operation
{
    NSPasteboard* pboard = [info draggingPasteboard];
    NSSet *allowedClasses = [NSSet setWithObjects: NSIndexSet.class, NSArray.class, NSURL.class, nil];
    NSError *error = nil;
    // internal drop track
    if ([[pboard types] containsObject:SBTracklistTableViewDataType] ) {
        NSData* rowData = [pboard dataForType:SBTracklistTableViewDataType];
        NSIndexSet* rowIndexes = [NSKeyedUnarchiver unarchivedObjectOfClasses: allowedClasses fromData: rowData error: &error];
        if (error != nil) {
            NSLog(@"Error unserializing index set %@", error);
            return NO;
        }
        NSMutableArray *tracks = [NSMutableArray array];
        NSArray *reversedArray  = nil;
        
        // get temp rows objects and remove them from the playlist
        [rowIndexes enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
            [tracks addObject:[[[SBPlayer sharedInstance] playlist] objectAtIndex:idx]];
            [[SBPlayer sharedInstance] removeTrackAtIndex: idx];
            [playlistTableView reloadData];
        }];
        
        // reverse track array
        reversedArray = [[tracks reverseObjectEnumerator] allObjects];
        
        // add reversed track at index
        for(SBTrack *track in reversedArray) {
            //NSLog(@"row : %ld", row);
            if(row > [[[SBPlayer sharedInstance] playlist] count])
                row--;
            
            [[SBPlayer sharedInstance] addTrack: track atIndex: row];
        }
        [playlistTableView reloadData];
        
    } else if([[pboard types] containsObject:SBLibraryTableViewDataType]) {
        
        NSData *data = [[info draggingPasteboard] dataForType:SBLibraryTableViewDataType];
        NSArray *tracksURIs = [NSKeyedUnarchiver unarchivedObjectOfClasses: allowedClasses fromData: data error: &error];
        if (error != nil) {
            NSLog(@"Error unserializing array %@", error);
            return NO;
        }
        
        // also add new track IDs to the array
        NSArray *reversedArray = [[tracksURIs reverseObjectEnumerator] allObjects];
        [reversedArray enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
            SBTrack *track = (SBTrack *)[self.managedObjectContext objectWithID:[[self.managedObjectContext persistentStoreCoordinator] managedObjectIDForURIRepresentation:obj]]; 
            
            [[SBPlayer sharedInstance] addTrack: track atIndex:row];
        }];
        
        
        [playlistTableView reloadData];
    }
    
    return YES;
}

#pragma mark -
#pragma mark UI Validation


- (BOOL)validateUserInterfaceItem: (id<NSValidatedUserInterfaceItem>) item {
    SEL action = item.action;
    
    BOOL tracksSelected = playlistTableView.selectedRow != -1;
    NSUInteger tracksSelectedCount = playlistTableView.numberOfSelectedRows;
    
    SBSelectedRowStatus selectedTrackRowStatus = 0;
    selectedTrackRowStatus = [self selectedRowStatus: [[SBPlayer sharedInstance] playlist] selectedIndices: playlistTableView.selectedRowIndexes];
    
    if (action == @selector(delete:)
        || action == @selector(createNewLocalPlaylistWithSelectedTracks:)) {
        return tracksSelected;
    }
    
    if (action == @selector(showSelectedInFinder:)) {
        return selectedTrackRowStatus & SBSelectedRowShowableInFinder;
    }
    
    if (action == @selector(downloadSelected:)) {
        return selectedTrackRowStatus & SBSelectedRowDownloadable;
    }
    
    if (action == @selector(showSelectedInLibrary:)) {
        return tracksSelectedCount == 1;
    }
    
    return true;
}

@end

File: Submariner/SBTracklistController.swift
//
//  SBTracklistController.swift
//  Submariner
//
//  Created by Calvin Buckley on 2024-01-31.
//
//  Copyright (c) 2024 Calvin Buckley
//  SPDX-License-Identifier: BSD-3-Clause
//  

import Cocoa

@objc class SBTracklistController: SBViewController, NSTableViewDelegate, NSTableViewDataSource, NSUserInterfaceValidations {
    @IBOutlet var playlistTableView: NSTableView!
    @IBOutlet var tracklistLengthView: NSTextField!
    
    @objc var databaseController: SBDatabaseController!
    
    private var notificationObserver: Any?
    
    override class func nibName() -> String! {
        "Tracklist"
    }
    
    override func loadView() {
        super.loadView()
        
        title = "Tracklist"
        
        playlistTableView.registerForDraggedTypes([.libraryItems, .libraryItem])
        
        notificationObserver = NotificationCenter.default.addObserver(forName: .SBPlayerPlaylistUpdated,
                                                                      object: nil,
                                                                      queue: nil,
                                                                      using: { notification in
            self.playlistTableView.reloadData()
        })
        
        tracklistLengthView.bind(.value,
                                 to: SBPlayer.sharedInstance(),
                                 withKeyPath: "playlist",
                                 options: [.valueTransformerName: "SBTrackListLengthTransformer"])
    }
    
    override var selectedTracks: [SBTrack]! {
        return SBPlayer.sharedInstance().playlist[playlistTableView.selectedRowIndexes]
    }
    
    // #MARK: - IBActions
    
    @IBAction func playSelected(_ sender: Any) {
        if playlistTableView.selectedRow != -1 {
            SBPlayer.sharedInstance().play(index: playlistTableView.selectedRow)
        }
    }
    
    @IBAction func delete(_ sender: Any) {
        if playlistTableView.selectedRow != -1 {
            SBPlayer.sharedInstance().remove(trackIndexSet: playlistTableView.selectedRowIndexes)
        }
    }
    
    @IBAction func cleanTracklist(_ sender: Any) {
        SBPlayer.sharedInstance().clear()
    }
    
    @IBAction func showSelectedInFinder(_ sender: Any) {
        if playlistTableView.selectedRow != -1 {
            self.showTracksInFinder(SBPlayer.sharedInstance().playlist, selectedIndices: playlistTableView.selectedRowIndexes)
        }
    }
    
    @IBAction func showSelectedInLibrary(_ sender: Any) {
        if playlistTableView.selectedRow != -1 {
            // only makes sense with a single track
            let track = SBPlayer.sharedInstance().playlist[playlistTableView.selectedRow]
            databaseController.go(to: track)
        }
    }
    
    @IBAction func downloadSelected(_ sender: Any) {
        if playlistTableView.selectedRow != -1 {
            self.downloadTracks(SBPlayer.sharedInstance().playlist, selectedIndices: playlistTableView.selectedRowIndexes, databaseController: databaseController)
        }
    }
    
    @IBAction func createNewLocalPlaylistWithSelectedTracks(_ sender: Any) {
        if playlistTableView.selectedRow != -1 {
            self.createLocalPlaylist(withSelected: SBPlayer.sharedInstance().playlist, selectedIndices: playlistTableView.selectedRowIndexes, databaseController: databaseController)
        }
    }
    
    // #MARK: - NSTableView DataSource
    
    func numberOfRows(in tableView: NSTableView) -> Int {
        return SBPlayer.sharedInstance().playlist.count
    }
    
    func tableView(_ tableView: NSTableView, objectValueFor tableColumn: NSTableColumn?, row: Int) -> Any? {
        // XXX: less ugly switch
        switch (tableColumn?.identifier.rawValue) {
        case "isPlaying" where row == SBPlayer.sharedInstance().currentIndex:
            return NSImage(systemSymbolName: "speaker.fill", accessibilityDescription: "Playing")
        case "title":
            return SBPlayer.sharedInstance().playlist[row].itemName
        case "artist":
            let track = SBPlayer.sharedInstance().playlist[row]
            if let artistName = track.artistName, artistName != "" {
                return artistName
            } else {
                return track.album?.artist?.itemName
            }
        case "duration":
            return SBPlayer.sharedInstance().playlist[row].durationString
        case "online":
            let track = SBPlayer.sharedInstance().playlist[row]
            if track.localTrack != nil || track.isLocal == true {
                return NSImage(systemSymbolName: "bolt.horizontal.fill", accessibilityDescription: "Cached")
            } else {
                return NSImage(systemSymbolName: "bolt.horizontal", accessibilityDescription: "Online")
            }
        default:
            return nil
        }
    }
    
    // #MARK: - NSTableView Delegate
    
    func tableView(_ tableView: NSTableView, pasteboardWriterForRow row: Int) -> (any NSPasteboardWriting)? {
        if tableView == playlistTableView {
            let track = SBPlayer.sharedInstance().playlist[row]
            return SBLibraryItemPasteboardWriter(item: track, index: row)
        }
        return nil
    }
    
    func tableView(_ tableView: NSTableView, validateDrop info: NSDraggingInfo, proposedRow row: Int, proposedDropOperation dropOperation: NSTableView.DropOperation) -> NSDragOperation {
        guard row != -1 && dropOperation == .above else {
            return []
        }
        
        if let sourceTable = info.draggingSource as? SBTableView, sourceTable == playlistTableView {
            return .move
        } else if info.draggingPasteboard.libraryItems() != nil {
            return .copy
        }
        return []
    }
    
    static let allowedClasses = [NSIndexSet.self, NSArray.self, NSURL.self]
    
    func tableView(_ tableView: NSTableView, acceptDrop info: NSDraggingInfo, row: Int, dropOperation: NSTableView.DropOperation) -> Bool {
        // XXX: For some reason, draggingSourceOperationMask has all bits set?
        if let sourceTable = info.draggingSource as? SBTableView, sourceTable == playlistTableView {
            let rowIndexes = info.draggingPasteboard.rowIndices()
            SBPlayer.sharedInstance().move(trackIndexSet: rowIndexes, index: row)
            
            // change selection to match new indices, since Array.move doesn't return them
            var newRow = row
            for index in rowIndexes {
                if index < newRow {
                    newRow -= 1
                }
            }
            let lastRow = newRow + rowIndexes.count - 1
            let newRange = newRow...lastRow
            let newIndexSet = IndexSet(integersIn: newRange)
            playlistTableView.selectRowIndexes(newIndexSet, byExtendingSelection: false)
        } else if let tracks = info.draggingPasteboard.libraryItems(managedObjectContext: self.managedObjectContext) {
            // handles both kinds of library track
            SBPlayer.sharedInstance().add(tracks: tracks, index: row)
        }
        
        return true
    }
    
    func tableViewSelectionDidChange(_ notification: Notification) {
        NotificationCenter.default.post(name: .SBTrackSelectionChanged, object: selectedTracks)
    }
    
    // #MARK: - UI Validator
    
    func validateUserInterfaceItem(_ item: NSValidatedUserInterfaceItem) -> Bool {
        let selectedTrackRowStatus = self.selectedRowStatus(SBPlayer.sharedInstance().playlist,
                                                            selectedIndices: playlistTableView.selectedRowIndexes)
        let count = playlistTableView.numberOfSelectedRows
        
        switch (item.action) {
        case #selector(SBTracklistController.downloadSelected(_:)):
            return selectedTrackRowStatus.contains(.downloadable)
        case #selector(SBTracklistController.showSelectedInFinder(_:)):
            return selectedTrackRowStatus.contains(.showableInFinder)
        case #selector(SBTracklistController.playSelected(_:)),
            #selector(SBTracklistController.delete(_:)),
            #selector(SBTracklistController.createNewLocalPlaylistWithSelectedTracks(_:)):
            return count > 0
        case #selector(SBTracklistController.showSelectedInLibrary(_:)):
            return count == 1
        default:
            return true
        }
    }
}
